File: Algorithms/mylibrary/src/main/java/it/car/tree/Trie.java
package it.car.tree;

/**
 * Created by ciriti on 21/08/16.
 */

public class Trie {

    public static final int ALPHABET_SIZE = 'z' - 'a';

    Node root = null;

    public static int charToIndex(char c){
        return 'z' - c;
    }

    public static class Node{
        Node[] children = new Node[ALPHABET_SIZE];
//        boolean isLeaf;
        char val;

        public Node(char val) {
            this.val = val;
        }
    }

    public Trie() {
        this.root = new Node('*');
    }

    public Node getRoot(){
        return  root;
    }

    public void insert(String s){
        recursiveInsert(root, s);
    }

    void recursiveInsert(Node n, String s){
        if(s.length() == 0){
            return;
        }

        char c = s.charAt(0);

        if( null == n.children[charToIndex(c)]){
            n.children[charToIndex(c)] = new Node(c);
        }

//        if(){
//
//        }
        recursiveInsert(n.children[charToIndex(c)], s.length() > 1 ? s.substring(1) : s);
    }

}

File: review_01_2018/src/main/java/com/ciriti/tree/Trie.kt
package com.ciriti.tree

/**
 * Created by Carmelo Iriti
 */

fun main(args: Array<String>) {
  val trie = Trie()
  trie.insert("carmelo")
  trie.insert("caramella")
  println(trie)
}

class Trie {
  private data class Node(
    val value: Char,
    val children: MutableMap<Char, Node> = mutableMapOf()
  )

  private val head = Node('-')

  fun insert(text: String) {
    var node = head
    text.forEach {
      node.children[it] ?: node.children.put(it, Node(it))
      node = node.children[it]!!
    }
  }

  fun searchAllWordsByPrefix(prefix: String): List<String> {

    val res = mutableListOf<String>()



    return emptyList()
  }

  private fun Node.dfs(
    index: Int,
    text: String,
    list: MutableList<String>
  ) {
  }

}